File: androidsupport/src/main/java/com/tq/libs/adapter/ArrayRecyclerAdapter.java
/*
 * ******************************************************************************
 *  Copyright Ⓒ 2016. TrinhQuan. All right reversed
 *  Author: TrinhQuan. Created on 2016/10/15
 *  Contact: trinhquan.171093@gmail.com
 * ******************************************************************************
 */

package com.tq.libs.adapter;

import android.content.Context;

import com.tq.libs.exception.RangeException;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.List;

public abstract class ArrayRecyclerAdapter<VH extends BaseRecyclerViewAdapter.BaseViewHolder<DATA>, DATA>
        extends BaseRecyclerViewAdapter<VH, DATA> {

    private List<DATA> mData = new ArrayList<>();

    public ArrayRecyclerAdapter(Context context) {
        super(context);
    }

    public void addData(DATA data) {
        int count = getItemCount();
        mData.add(data);
        notifyItemInserted(count);
    }

    public void addData(Collection<DATA> datas) {
        int original = getItemCount();
        mData.addAll(datas);
        notifyItemRangeInserted(original, datas.size());
    }

    public void setData(DATA data) {
        clear();
        addData(data);
    }

    public void setData(Collection<DATA> datas) {
        clear();
        addData(datas);
    }

    private void clear() {
        int count = getItemCount();
        mData.clear();
        notifyItemRangeRemoved(0, count);
    }

    public <T extends List<DATA>> T getData(T des) {
        Collections.copy(des, mData);
        return des;
    }

    public List<DATA> getData() {
        List<DATA> datas = new ArrayList<>(getItemCount());
        return getData(datas);
    }

    public List<DATA> getDataByReference() {
        return mData;
    }

    @Override
    public void release() {
        super.release();
        mData.clear();
    }

    @Override
    public int getItemCount() {
        return mData.size();
    }

    public DATA getItemAtPosition(int position) {
        if (position < 0 || position >= getItemCount()) {
            throw new RangeException();
        }
        return mData.get(position);
    }
}

File: kotlinandroidsupport/src/main/java/com/tq/klibs/adapter/ArrayRecyclerAdapter.kt
/*
 * ******************************************************************************
 *  Copyright Ⓒ 2016. TrinhQuan. All right reserved
 *  Author: TrinhQuan. Created on 2016/12/20
 *  Contact: trinhquan.171093@gmail.com
 * ******************************************************************************
 */

package com.tq.klibs.adapter

import android.content.Context
import com.tq.klibs.exception.RangeException
import java.util.*

abstract class ArrayRecyclerAdapter<VH, DATA>(context: Context?)
: BaseRecyclerViewAdapter<VH, DATA>(context)
where VH : BaseRecyclerViewAdapter.BaseViewHolder<DATA> {

    private val mData = ArrayList<DATA>()

    open fun addData(data: DATA) {
        val count = itemCount
        mData.add(data)
        notifyItemInserted(count)
    }

    open fun addData(data: Collection<DATA>) {
        val original = itemCount
        mData.addAll(data)
        notifyItemRangeInserted(original, data.size)
    }

    open fun setData(data: DATA) {
        clear()
        addData(data)
    }

    open fun setData(datas: Collection<DATA>) {
        clear()
        addData(datas)
    }

    private fun clear() {
        val count = itemCount
        mData.clear()
        notifyItemRangeRemoved(0, count)
    }

    fun <T : MutableList<DATA>> getData(des: T): T {
        Collections.copy(des, mData)
        return des
    }

    val data: MutableList<DATA>
        get() {
            return getData<MutableList<DATA>>(ArrayList<DATA>(itemCount))
        }

    val dataByReference: MutableList<DATA>
        get() = mData

    override fun release() {
        super.release()
        mData.clear()
    }

    override fun getItemCount(): Int {
        return mData.size
    }

    override fun getItemAtPosition(position: Int): DATA? {
        if (position < 0 || position >= itemCount) {
            throw RangeException()
        }
        return mData[position]
    }
}
